### PR TITLE
Explosives - Add ability to attach explosives to objects

### PR DESCRIPTION
**Pull request by DartRuffian**
**When merged this pull request will:**
- Add a config flag to allow explosives to be placed on any non-unit object (e.g. houses)
  - M112 and the M183 can be attached to objects (vibes based, maybe not the M183 though)
  - This does have issues where units can place/attach explosives inside thin walls (like doors), but this also happens with vehicles anyway
- Add common function to check if something is a terrain object 

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/explosives/functions/fnc_setupExplosive.sqf [hunk 1]
```diff
@@ -57,6 +57,7 @@ _unit setVariable [QGVAR(cancelActionEH), [_unit, "zoomtemp", {true}, {GVAR(plac
 private _supportedTriggers = getArray (_configMagazine >> "ACE_Triggers" >> "SupportedTriggers");
 private _aceTriggers = configFile >> "ACE_Triggers";
 private _isAttachable = _supportedTriggers findIf {(getNumber (_aceTriggers >> _x >> "isAttachable")) == 1} != -1;
+private _isObjectAttachable = getNumber (_configMagazine >> QGVAR(isObjectAttachable)) == 1;
```
Comment (DartRuffian) [inline addons/explosives/functions/fnc_setupExplosive.sqf]: I couldn't think of a better name :P
Lmk if someone's got something better

**Comment by DartRuffian** [comment]
I was gonna do the SLAM but it's rotated 90 degrees so I just left it off

File: addons/explosives/functions/fnc_setupExplosive.sqf (hunk 1 above)
Comment (Dystopian) [reply addons/explosives/functions/fnc_setupExplosive.sqf]: `isAttachableToObject` or `isAttachableTo`?

Comment (DartRuffian) [reply addons/explosives/functions/fnc_setupExplosive.sqf]: First seems better, I wanted to be clear that its specific to objects since most things can already be attached to vehicles

**Comment by DartRuffian** [comment]
~~Found an issue while doing some more testing, seems like trying to attach them to terrain objects causes them to be sent to the debug corner~~ Resolved by using a dummy object when attaching to terrain objects


https://github.com/user-attachments/assets/3585a4b1-429a-4d10-a0c9-2a5d7ebac0a0

File: addons/explosives/functions/fnc_setupExplosive.sqf (hunk 1 above)
Comment (DartRuffian) [reply addons/explosives/functions/fnc_setupExplosive.sqf]: Could just name it isSticky

**Comment by DartRuffian** [comment]
~~Needs docs, but that's easy. Can do that after property name is picked~~ Done

**Comment by DartRuffian** [comment]
Name was changed to make it clearer, I'm happy with it now

File: addons/common/functions/fnc_isTerrainObject.sqf
Comment (johnb432) [inline addons/common/functions/fnc_isTerrainObject.sqf:1]: Is there a reason not to use https://cbateam.github.io/CBA_A3/docs/files/common/fnc_isTerrainObject-sqf.html instead? I'd prefer to not add a public function if we can avoid it.

File: addons/explosives/functions/fnc_setupExplosive.sqf [hunk 2]
```diff
@@ -207,7 +213,17 @@ GVAR(TweakedAngle) = 0;
...
                 _placeAngle = _placeAngle + 180; // CfgAmmos seem to be 180 for some reason
             } else {
-                private _modelOffset = _attachVehicle worldToModel (_virtualPosASL call EFUNC(common,ASLToPosition));
+                private _modelOffset = [0, 0, 0];
+
+                // Terrain objects (seemingly) can't have objects attached to them, so create a dummy instead
+                if (_attachVehicle call EFUNC(common,isTerrainObject)) then {
+                    _attachVehicle = createVehicle ["Helper_Base_F", [0, 0, 0], [], 0, "CAN_COLLIDE"];
+                    _attachVehicle setPosASL _virtualPosASL;
+                    _attachVehicle setDir _placeAngle;
```
Comment (johnb432) [inline addons/explosives/functions/fnc_setupExplosive.sqf]:
```suggestion
                // Terrain objects (seemingly) can't have objects attached to them, so create a dummy instead
                private _modelOffset = if (_attachVehicle call EFUNC(common,isTerrainObject)) then {
                    _attachVehicle = createVehicle ["Helper_Base_F", [0, 0, 0], [], 0, "CAN_COLLIDE"];
                    _attachVehicle setPosASL _virtualPosASL;
                    _attachVehicle setDir _placeAngle;
                    
                    [0, 0, 0]
```

File: addons/explosives/functions/fnc_setupExplosive.sqf [hunk 3]
```diff
@@ -121,14 +122,19 @@ GVAR(TweakedAngle) = 0;
         } forEach [[0, 0], [-TEST_LENGTH, -TEST_LENGTH], [TEST_LENGTH, -TEST_LENGTH], [-TEST_LENGTH, TEST_LENGTH], [TEST_LENGTH, TEST_LENGTH]];
 
         if (
-            !isNull _attachVehicle &&
-            {(_attachVehicle isKindOf "Car") || {_attachVehicle isKindOf "Tank"} || {_attachVehicle isKindOf "Air"} || {_attachVehicle isKindOf "Ship"}} &&
+            !isNull _attachVehicle && (
+                // Allow attaching to non-unit objects (if set) or any vehicle
+                (_isSticky && {!(_attachVehicle isKindOf "CAManBase")}) || {(_attachVehicle isKindOf "Car")} ||
```
Comment (johnb432) [inline addons/explosives/functions/fnc_setupExplosive.sqf]: I don't think we want to allow attaching explosives to animals.
```suggestion
                (_isSticky && {!(_attachVehicle isKindOf "Man")}) ||
                {_attachVehicle isKindOf "Car"} ||
```

File: addons/common/functions/fnc_isTerrainObject.sqf
Comment (DartRuffian) [reply addons/common/functions/fnc_isTerrainObject.sqf:1]: I didn't know it existed, and seemed like common enough utility to just make it a public function

Comment (DartRuffian) [reply addons/common/functions/fnc_isTerrainObject.sqf:1]: Just checked the CBA one, it's doing a nearestTerrainObjects so its decently slower than just checking net id:
```sqf
player call CBA_fnc_isTerrainObject; // Execution Time: 0.0041 ms  |  Cycles: 10000/10000  |  Total Time: 41 ms
player call ace_common_fnc_isTerrainObject; // Execution Time: 0.0037 ms  |  Cycles: 10000/10000  |  Total Time: 37 ms
```

I'll swap it to the CBA one, and PR changing the CBA one to check net id

File: addons/explosives/functions/fnc_setupExplosive.sqf (hunk 3 above)
Comment (DartRuffian) [reply addons/explosives/functions/fnc_setupExplosive.sqf]: Don't ruin my plans

Image

Comment (SpicyBagpipes) [reply addons/explosives/functions/fnc_setupExplosive.sqf]: Soviet anti tank dogs

File: addons/common/functions/fnc_isTerrainObject.sqf
Comment (DartRuffian) [reply addons/common/functions/fnc_isTerrainObject.sqf:1]: https://github.com/CBATeam/CBA_A3/pull/1816
